Preserve other existing IP configurations when apply()

cathay4t opened the pull request:
* With this fix `netapplier.apply()` will not discard all IP configurations(DNS, route, etc) but keep existing  one except IP addresses.

~~This PR is based on #107 , please review it first.~~

tyll commented on libnmstate/netapplier.py:
```diff
@@ -313,12 +315,20 @@ def assert_ifaces_state(ifaces_desired_state, ifaces_current_state):
                 'desired: {}\ncurrent: {}'.format(iface_dstate, iface_cstate))
 
 
+def _sort_ip_addresses(iface):
```
I remember that @thom311 told me that the order of IP addresses matters but I forgot for what exactly. Maybe for the default IP address that is used for outgoing connections via that interface.

cathay4t commented:
Fixing unit test. Please hold the review.

cathay4t commented:
Unit test fixed. OK for review.

tyll commented on libnmstate/nm/ipv4.py:
```diff
@@ -20,8 +20,17 @@
 
-def create_setting(config):
-    setting_ipv4 = nmclient.NM.SettingIP4Config.new()
+def create_setting(config, base_remote_conn):
```
Please do change `base_remote_conn` to `base_con_profile` to have it uniform in nmstate (also  for IPv6).

cathay4t replied:
Done. Thanks.

tyll commented:
@EdDev stdci failed with:
```
+ docker_exec 'systemctl start dbus.socket'

+ docker exec -i c9b16afaff56db211391b0fc0b7944308223a57a524ad96b9b362ae37ba2e318 /bin/bash -c 'systemctl start dbus.socket'

Failed to get D-Bus connection: No such file or directory

+ remove_container

+ res=1

+ '[' 1 -ne 0 ']'

+ echo '*** ERROR: 1'

```
http://jenkins.ovirt.org/job/nmstate_nmstate_standard-check-pr/182/display/redirect

tyll commented on tests/lib/nm/ipv4_test.py:
```diff
@@ -29,15 +29,16 @@ def NM_mock():
 
 def test_create_setting_without_config(NM_mock):
-    ipv4_setting = nm.ipv4.create_setting(config=None)
+    ipv4_setting = nm.ipv4.create_setting(config=None, base_remote_conn=None)
```
this also needs to be updated to base_con_profile and the other occurences (see the travis failure) or run tox locally.

EdDev commented on libnmstate/nm/ipv4.py:
```diff
@@ -33,6 +42,7 @@ def create_setting(config):
 
 
 def _add_addresses(setting_ipv4, addresses):
+    setting_ipv4.clear_addresses()
```
Why it is done twice? (line 29 does it as well)

cathay4t replied:
Just some leftover. Fixed.

EdDev commented on tests/lib/nm/ipv4_test.py:
```diff
@@ -48,7 +49,8 @@ def test_create_setting_without_addresses(NM_mock):
             'address': [],
-        }
+        },
+        base_con_profile=None
```
What is this?

cathay4t replied:
It's `nm.ipv4.create_setting(config={ 'enabled': True, 'address': [],}, base_con_profile=None)`

EdDev commented on tests/lib/nm/ipv6_test.py:
```diff
@@ -48,7 +49,8 @@ def test_create_setting_without_addresses(NM_mock):
             'address': [],
-        }
+        },
+        base_con_profile=None
```
What is this?

cathay4t replied:
It's `nm.ipv6.create_setting(config={ 'enabled': True, 'address': [],}, base_con_profile=None)`

cathay4t commented:
@EdDev New patch uploaded based on your feedback. Thank you.

cathay4t commented:
Both unit test and integration test has been attached.

This PR is based on PR #107 in order to pass the integration test.

EdDev commented on libnmstate/netapplier.py:
```diff
@@ -341,3 +344,11 @@ def _cleanup_iface_ipv6_link_local_addr(desired_state, current_state):
                 state['ipv6']['address'].remove(addr_dict)
+
+
+def _sort_ip_addresses(desired_state, current_state):
```
It will fit better outside this module, perhaps to an ipaddress module. We can delay this to an overall cleanup of netapplier.

EdDev commented on tests/integration/reapply_test.py:
```diff
@@ -0,0 +1,140 @@
+    return active_conn.get_uuid()
+
+
+def test_reapply_preserve_ip_config():
```
It have trouble understanding the test, it feels too complex to me.
After you apply the base config (and knowing already the device manipulated), you can just execute `nmcli con modify` directly to add a special extra parameter. No need for a generic dict with these embedded.

In addition, the IPv4 and IPv6 scenarios are different, they should have their own test separately.

You also have libcmd.exec_cmd in you disposal, check the nmstatectl_show_test test module.
You can just check that an extra config value is still there by a simple `a in b` assert.

cathay4t commented on tests/integration/reapply_test.py:
```diff
@@ -0,0 +1,140 @@
+from subprocess import check_output
+from subprocess import check_call
+
+import gi                           # pylint: disable=import-error
```
I through we should not use the code we are testing against and nmclient seems not in the
part of public available interface of libnmstate.

cathay4t commented on tests/integration/reapply_test.py:
```diff
@@ -0,0 +1,140 @@
+
+
+
+def _get_nm_profile_uuid(iface_name):
```
Like previous comments: Should we expose functions via libnmstate for integration test seek?

cathay4t commented:
V2 patch uploaded.

Changes since V1:
 * Squashed unit test and integration test into production code patch.
 * Use `testlib.cmd` instead of `subprocess.check_output()` and `subprocess.check_call()`.
 * Trivial fixes suggested above.

cathay4t commented:
Rebased to latest master.

EdDev commented on tests/integration/preserve_ip_config_test.py:
```diff
@@ -0,0 +1,136 @@
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+
+import gi                           # pylint: disable=import-error
```
All these are already available in the nm.nmclient module, you can use it.

EdDev commented on tests/integration/preserve_ip_config_test.py:
```diff
@@ -0,0 +1,136 @@
+    }
+]
+
+# TEST-NET addresses: https://tools.ietf.org/html/rfc5737#section-3
```
I think we do not need to reason on why a specific address is used.
If it is important for you to mention it, you can do so in the commit message.

EdDev commented on tests/integration/preserve_ip_config_test.py:
```diff
@@ -0,0 +1,136 @@
+IPV6_ADDRESS1 = '2001:db8:1::1'
+
+
+def _get_nm_profile_uuid(iface_name):
```
Please use nm.device module for this. You have there `get_device_by_name` and `list_devices`.
Probably you actually just need `connection.get_device_active_connection`.

If you decide to keep this helper, please locate it below the callers.

EdDev commented on tests/integration/preserve_ip_config_test.py:
```diff
@@ -0,0 +1,136 @@
+    uuid = _get_nm_profile_uuid(iface_name)
+    old_cfg = _get_cur_extra_ip_config(uuid)
+
+    try:
```
Please use a contextmanager for the `_apply_extra_ip_config`.
```
with _extra_ip_config():
    netapplier.apply(cur_state)
    _assert_extra_ip_config(...)
```

EdDev commented on tests/integration/preserve_ip_config_test.py:
```diff
@@ -0,0 +1,136 @@
+                                       ','.join(props), 'connection', 'show',
+                                       uuid])
+    assert rc == 0
+    return [
```
Why do we need this structure? It would be less messy with a direct approach of fetching and setting individual properties.

cathay4t replied:
I use that to avoid using magic string for the property name `ipv4.dad-timeout` and `ipv6.dhcp-hostname`. Do you have better approach does not require hard code magic string?

EdDev replied:
I think that packaging the two in a dict structure is not serving anything. Each should be self contained not coupled together.

cathay4t commented:
V3 patch uploaded.

Changes since V2:
 * Moved the `preserve_ip_config_test.py` integration test back to `tests/integration` folder.
 * Replaced the complex structure for saving extra configuration key and value to `tuple`.

cathay4t commented on tests/integration/preserve_ip_config_test.py:
```diff
@@ -0,0 +1,122 @@
+IPV6_ADDRESS1 = '2001:db8:1::1'
+
+
+def test_reapply_preserve_ip_config():
```
Oh, I missed that.

cathay4t commented:
V4 patch uploaded.

Changes since V3:
 * Fixed the integration test based on feedback.

EdDev commented:
Thank you!
